### ISSUE TITLE
Replace likely macro with C++20 attribute

### ISSUE THREAD

**Issue by heinezen** [SFTtech/openage#1514]
We currently use gcc's `__builtin_expect()` macro aliased as `unlikely` and `likely` for static branch prediction. Primarily, this optimizes checks for parser errors. This optimizes performance for code paths which are only taken under very specific circumstances.

Example:

```cpp
if (unlikely(not this->value.exists())) {
    throw InternalError{"fetched nonexisting value of member"};
}
```

Since C++20, the standard supports [attributes](https://en.cppreference.com/w/cpp/language/attributes/likely) that do the same thing. Additionally, they are compiler agnostic, so they should also work in clang or msvc.  They should also be more readable since they don't directly wrap the condition.

Basically the code gets changed to this:

```cpp
if (not this->value.exists()) [[unlikely]] {
    throw InternalError{"fetched nonexisting value of member"};
}
```

If you want to discover the codebase or really like mundane tasks, then this is an issue for you :)

----
Also take note of the same issue in the nyan repository: https://github.com/SFTtech/nyan/issues/105

**Comment by tarunsamanta2k20** [comment]
I'd like to work on it.

**Comment by heinezen** [comment]
@tarunsamanta2k20 Sweet :) Tell us if you need any directions.

**Comment by derekfrogget** [comment]
@tarunsamanta2k20 Are you still working on this? I would like to work on it if you are not doing it anymore.

**Comment by tarunsamanta2k20** [comment]
@derekfrogget let me do by sometime.

**Comment by tarunsamanta2k20** [comment]
@heinezen should i replace all occurrences of likely and unlikely both?

**Comment by tarunsamanta2k20** [comment]
@heinezen if you want you may do it for '**likely**' only. I am doing this only for unlikely cases.

**Comment by heinezen** [comment]
@tarunsamanta2k20 @derekfrogget both likely and unlikel can be replaced.

There's also the same issue for the nyan repo https://github.com/SFTtech/nyan/issues/105 (i think  you also wanted to do this @tarunsamanta2k20 , right?)

**Comment by heinezen** [comment]
now it's only `likely` thats still there

**Comment by tarunsamanta2k20** [comment]
> @tarunsamanta2k20 @derekfrogget both likely and unlikel can be replaced.
> 
> There's also the same issue for the nyan repo [SFTtech/nyan#105](https://github.com/SFTtech/nyan/issues/105) (i think you also wanted to do this @tarunsamanta2k20 , right?)

@derekfrogget will do for likely. Ping him.

**Comment by heinezen** [comment]
Thanks everybody, now only the nyan repo task is left https://github.com/SFTtech/nyan/issues/105 :)
